controllers/helmrelease_controller.go

```diff
@@ -499,6 +501,57 @@ func (r *HelmReleaseReconciler) getServiceAccountToken(ctx context.Context, hr v
 	return token, nil
 }
 
+const (
+	// Annotation on resources to explicitly share them with other
+	// namespaces. The annotation must be either an array of strings, containing the
+	// names of the namespaces with which the resource is shared or a string with
+	// the wildcard value "*" to share the resource with all namesapces.
+	shareWithAnnotation = "helm.toolkit.fluxcd.io/share-with"
+)
+
+// Check if a resource is shared with a given namespace, according to the following rules:
+//
+// 1. If resource is in the same namespace, it is shared by default and can be accessed.
+// 2. If resource is in a different namespace and has a "helm.toolkit.fluxcd.io/share-with"
+//    annotation of type string with the wildcard value "*", then it is shared and
+//    can be accessed.
+// 3. If resource is in a different namespace and has a "helm.toolkit.fluxcd.io/share-with"
+//    annotation (an array of namespace names) which contains the namespace, it is
+//    shared and can be accessed.
+// 4. If non of the above rules applies, the resource is not shared and cannot be
+//    accessed.
+func isSharedWith(kind string, resource *metav1.ObjectMeta, namespace string) error {
+	if resource != nil {
+		if resource.Namespace == namespace {
+			// Rule 1.
+			return nil
+		}
+		if resource.Annotations != nil {
+			shareWithString := strings.TrimSpace(resource.Annotations[shareWithAnnotation])
+			if shareWithString == "*" {
+				// Rule 2.
+				return nil
+			}
+			if shareWithString != "" {
+				var shareWith []string = make([]string, 0)
+				if err := json.Unmarshal([]byte(shareWithString), &shareWith); err != nil {
+					return fmt.Errorf(
+						"%s '%s/%s' has invalid %s annotation value, expected string or array of string got '%s': %w",
+						kind, resource.Namespace, resource.Name, shareWithAnnotation, shareWithString, err)
+				}
+				if slice.ContainsString(shareWith, namespace, nil) {
+					// Rule 3.
+					return nil
+				}
+			}
+		}
+	}
+	// Rule 4.
+	return fmt.Errorf(
+		"%s '%s/%s' is not shared with namespace '%s' (change its '%s' annotation to enable sharing)",
+		kind, resource.Namespace, resource.Name, namespace, shareWithAnnotation)
```

Comment:
Consider exiting early to reduce nested logic.
No need to test a `map[string]string` for nil before checking if it has a particular key, use `ok` as demonstrated.
I took out checking for an empty string as that will fail unmarshalling which I felt simplified things, but that could be personal preference.
metav1.ResourceObj keys should use the Getters when available, ie. GetAnnotations()
```suggestion
	if resource == nil {
		return fmt.Error("nil resource")
	}
	shareWithString, ok := resource.GetAnnotations()[shareWithAnnotation]
	if !ok {
		// Rule 4.
		return fmt.Errorf(
			"%s '%s/%s' is not shared with namespace '%s' (add an annotation, '%s', to enable sharing)",
			kind, resource.Namespace, resource.Name, namespace, shareWithAnnotation)
	}
	shareWithString = strings.TrimSpace(shareWithString)
	if resource.Namespace == namespace {
		// Rule 1.
		return nil
	}
	if shareWithString == "*" {
		// Rule 2.
		return nil
	}
	var shareWith []string
	if err := json.Unmarshal([]byte(shareWithString), &shareWith); err != nil {
		return fmt.Errorf(
			"%s '%s/%s' has invalid %s annotation value, expected string or array of string got '%s': %w",
			kind, resource.Namespace, resource.Name, shareWithAnnotation, shareWithString, err)
	}
	if !slice.ContainsString(shareWith, namespace, nil) {
		// Rule 4.
		return fmt.Errorf(
			"%s '%s/%s' is not shared with namespace '%s' (change its '%s' annotation to enable sharing)",
			kind, resource.Namespace, resource.Name, namespace, shareWithAnnotation)
	}
	// Rule 3.
	return nil
```

I hope this is helpful.